### PR TITLE
Retire Federated Runtime

### PR DESCRIPTION
**Pull request by KPrasch**
**Type of PR:**
Removal

**Required reviews:** 
3

**What this does:**
- Deprecates federated mode runtime (full stack)
- De-duplicates federated/decentralized tests
- Adapts all federated tests and fixtures to run in "decentralized" mode
- Removes federated demos
- Removes federated test/ops automation
- "Reinvention" of federated mode by way of mocking

**Issues fixed/closed:**
- supports closure of https://github.com/nucypher/tdec/issues/80

**Why it's needed:**
- Federated mode is unused outside of tests and is one of the largest contributors to cognitive overhead in the codebase.  
- It's original intention and purpose (early prototyping of off-chain distributed PRE) has been fulfilled.  
- In the current era of the nucypher network it can safely be retired. 
- Substantially fewer logical branches, de-duplication, simplified function signatures, and simplification of complex polymorphic patterns.

**Notes for reviewers:**
- This PR is a draft including the commit history which will be revised.
- The `--dev` flag needs to be considered for removal, as it's unclear what the intended behavior is without federated mode.
- Several tests are marked as `skip` in this PR that need debugging:
  - [x] `test_ursula_startup_ip_checkup`
  - [x] `test_ursula_run_ip_checkup`
  - [x] `test_alice_can_learn_about_a_whole_bunch_of_ursulas`
  - [x] `test_alice_verifies_ursula_just_in_time`
  - [x] `test_use_external_cache`
  
 - Co-Author: @jMyles

### CODE REVIEW

File: nucypher/network/protocols.py [hunk 1]
```diff
@@ -9,17 +11,16 @@ class SuspiciousActivity(RuntimeError):
     """raised when an action appears to amount to malicious conduct."""
 
 
-def parse_node_uri(uri: str):
+def parse_node_uri(uri: str, delimiter: str = '@') -> Tuple[str, int, Address]:
```
Comment (derekpierre) [inline nucypher/network/protocols.py:14]: Nice!

File: nucypher/network/protocols.py [hunk 2]
```diff
@@ -9,17 +11,16 @@ class SuspiciousActivity(RuntimeError):
...
-    if '@' in uri:
-        checksum_address, uri = uri.split("@")
+    checksum_address = None
+    if delimiter in uri:
+        checksum_address, uri = uri.split(delimiter)
         if checksum_address is None:
             raise ValueError(f"{uri} is not a valid Teacher URI - no checksum address.")
         if not is_checksum_address(checksum_address):
             raise ValueError("{} is not a valid checksum address.".format(checksum_address))
-    else:
```
Comment (derekpierre) [inline nucypher/network/protocols.py:21]: Should we raise a value error if delimiter not in uri?

File: nucypher/policy/reservoir.py [hunk 3]
```diff
@@ -1,36 +1,19 @@
...
 
-def make_decentralized_staking_provider_reservoir(application_agent: PREApplicationAgent,
-                                                  exclude_addresses: Optional[Iterable[ChecksumAddress]] = None,
-                                                  include_addresses: Optional[Iterable[ChecksumAddress]] = None,
-                                                  pagination_size: int = None):
+def make_staking_provider_reservoir(
+    application_agent: PREApplicationAgent,
+    exclude_addresses: Optional[Iterable[ChecksumAddress]] = None,
+    include_addresses: Optional[Iterable[ChecksumAddress]] = None,
+    pagination_size: int = None,
```
Comment (derekpierre) [inline nucypher/policy/reservoir.py:15]:
```suggestion
    pagination_size: Optional[int] = None,
```

**Review by derekpierre — COMMENTED** [review]
Good stuff! Still working my way through

File: nucypher/characters/lawful.py [hunk 4]
```diff
@@ -566,115 +536,88 @@ class NotEnoughUrsulas(Learner.NotEnoughTeachers):
...
             self._metadata = None
 
-            # Operating Mode
-            self.known_node_class.set_federated_mode(federated_only)
-
             # Health Checks
             self._availability_check = availability_check
             self._availability_tracker = AvailabilityTracker(ursula=self)
-            if not federated_only:
-                self._operator_bonded_tracker = OperatorBondedTracker(ursula=self)
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/lawful.py:627]: I relocated it to `Operator.__init__` in the abstraction leak commit.

File: tests/unit/test_bytestring_types.py [hunk 5]
```diff
@@ -5,12 +5,12 @@
...
 
 
-def test_message_kit_serialization_via_enrico(federated_alice):
+def test_message_kit_serialization_via_enrico(alice):
 
     mock_label = b'this is a label'
 
     # Enrico
-    enrico = Enrico.from_alice(federated_alice, label=mock_label)
+    enrico = Enrico.from_alice(alice, label=mock_label)
```
Comment (derekpierre) [inline tests/unit/test_bytestring_types.py:13]: We can modify this unit test to not take the `alice` fixture and therefore not need testerchain/any blockchain ursula creation.

```suggestion
    enrico = Enrico(policy_encrypting_key=SecretKey.random().public_key())
```

Then the `alice` fixture is not needed.

File: tests/unit/test_character_serialization.py [hunk 6]
```diff
@@ -1,10 +1,8 @@
-
-
 from nucypher.characters.lawful import Ursula
 
 
-def test_serialize_ursula(federated_ursulas):
-    ursula = federated_ursulas.pop()
+def test_serialize_ursula(ursulas):
```
Comment (derekpierre) [inline tests/unit/test_character_serialization.py:4]: If we modify this test to just create a simple Ursula OR move this test to integration, then the timing for `tests/unit` tests goes down from ~23s to 3s.

File: tests/acceptance/cli/ursula/test_ursula_cli_ip_detection.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+from nucypher.utilities.networking import UnknownIPAddress
+from tests.constants import (
+    FAKE_PASSWORD_CONFIRMED,
+    MOCK_IP_ADDRESS,
+    TEST_ETH_PROVIDER_URI,
+    TEST_POLYGON_PROVIDER_URI,
+)
+
+
+@pytest.mark.skip("Hangs for unknown reason.  Refuses to exit.")
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply tests/acceptance/cli/ursula/test_ursula_cli_ip_detection.py:16]: I suspect the test is prompting and waiting for an `operator address` (no more of this federated business :swe. You may need to either provide it as a CLI argument, or select an index via `user_input`.

File: tests/unit/conftest.py [hunk 8]
```diff
@@ -10,3 +16,45 @@ def mock_ethereum_client(mocker):
     web3_mock = mocker.Mock()
     mock_client = MockEthereumClient(w3=web3_mock)
     return mock_client
+
+
+@pytest.fixture(scope='module')
+def random_account():
+    key = Account.create(extra_entropy="lamborghini mercy")
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/unit/conftest.py:23]: Don't look at me this was GitHub co-pilot's idea!

File: nucypher/policy/reservoir.py (hunk 3 above)
Comment (KPrasch) [reply nucypher/policy/reservoir.py:15]: 👍🏻

File: nucypher/network/protocols.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/network/protocols.py:21]: No, because the delimiters presence is optional and this function parses URLs with or without it.  I'm unsure if it's being used anymore, but theoretically peers can be verified and filtrated by a checksum address.

File: nucypher/cli/painting/nodes.py [hunk 9]
```diff
@@ -32,16 +32,14 @@ def paint_node_status(emitter, ursula, start_time):
...
              'Node Storage Type ... {}'.format(ursula.node_storage._name.capitalize()),
              'Known Nodes ......... {}'.format(len(ursula.known_nodes)),
              teacher]
 
-    if not ursula.federated_only:
-        operator_address = 'Operator Address ...... {}'.format(ursula.operator_address)
-        current_period = f'Current Period ...... {ursula.application_agent.get_current_period()}'
-        stats.extend([current_period, operator_address])
+    operator_address = 'Operator Address ...... {}'.format(ursula.operator_address)
+    current_period = f'Current Period ...... {ursula.application_agent.get_current_period()}'
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/painting/nodes.py:41]: 👍🏻 removed

File: nucypher/characters/lawful.py [hunk 10]
```diff
@@ -1042,7 +982,7 @@ def from_seed_and_stake_info(cls,
         )
 
         # Check the node's stake (optional)
-        if minimum_stake > 0 and staking_provider_address and not federated_only:
+        if minimum_stake > 0 and staking_provider_address:
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/lawful.py:985]: It's interesting functionality and does seem more relevant without federated mode.  In any case I'd prefer to leave that for a different PR.

File: nucypher/characters/lawful.py [hunk 11]
```diff
@@ -143,33 +135,28 @@ def __init__(self,
...
                            known_node_class=Ursula,
                            is_me=is_me,
-                           federated_only=federated_only,
                            eth_provider_uri=eth_provider_uri,
                            checksum_address=checksum_address,
                            network_middleware=network_middleware,
                            *args, **kwargs)
 
-        if is_me and not federated_only:  # TODO: #289
-            if not eth_provider_uri:
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/lawful.py:153]: Conceptually the only reason to use optional parameters at all is to support the instantiation of peers using the same object.  This particular requirement is relocated to `Operator.__init__`.  Ideally the abstraction layers between `Character` and `Operator` can be fixed after this PR but it's a bit sensitive.

File: tests/acceptance/cli/ursula/test_ursula_cli_ip_detection.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/acceptance/cli/ursula/test_ursula_cli_ip_detection.py]: fixed

File: tests/unit/test_character_serialization.py (hunk 6 above)
Comment (derekpierre) [reply tests/unit/test_character_serialization.py]: 🚀 👌

File: tests/unit/conftest.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (derekpierre) [reply tests/unit/conftest.py]: hehe, it's a rap song reference - https://www.youtube.com/watch?v=7Dqgr0wNyPo&ab_channel=KanyeWestVEVO

File: nucypher/cli/commands/ursula.py [hunk 12]
```diff
@@ -170,8 +169,8 @@ def generate_config(self, emitter, config_root, force, key_material):
         if self.dev:
             raise RuntimeError('Persistent configurations cannot be created in development mode.')
 
-        if (not self.operator_address) and not self.federated_only:
-            prompt = "Select operator account"
+        if not self.operator_address:
```
Comment (derekpierre) [inline nucypher/cli/commands/ursula.py:172]: Should `force` be considered here and other places with undefined information i.e. raise and don't prompt if force is true?

File: tests/unit/test_character_serialization.py [hunk 13]
```diff
@@ -1,11 +1,8 @@
-
-
 from nucypher.characters.lawful import Ursula
 
 
-def test_serialize_ursula(federated_ursulas):
-    ursula = federated_ursulas.pop()
-    ursula_as_bytes = bytes(ursula.metadata())
+def test_serialize_ursula(light_ursula):
```
Comment (derekpierre) [inline tests/unit/test_character_serialization.py:4]: 💥

File: tests/acceptance/cli/test_status.py [hunk 14]
```diff
@@ -1,115 +0,0 @@
...
-    for agent in agents:
-        contract_regex = f"^{agent.contract_name} \\.+ {agent.contract_address}"
-        assert re.search(contract_regex, result.output, re.MULTILINE)
-
-    assert re.search(f"^Provider URI \\.+ {TEST_ETH_PROVIDER_URI}", result.output, re.MULTILINE)
-    assert re.search(f"^Current Period \\.+ {staking_agent.get_current_period()}", result.output, re.MULTILINE)
-
-
-@pytest.mark.skip()
-def test_nucypher_status_events(click_runner, testerchain, agency_local_registry, staking_providers, temp_dir_path):
```
Comment (derekpierre) [inline tests/acceptance/cli/test_status.py:45]: This test might still be applicable...

Comment (KPrasch) [reply tests/acceptance/cli/test_status.py:45]: Is this command being used?  This test has been skipped for quite some time and there are name errors inside.  Admittedly this removal is a little bit out of scope.

Comment (derekpierre) [reply tests/acceptance/cli/test_status.py:45]: Ah right, I missed that the test was skipped. I see now. The issue is that the status cli retrieves legacy StakingEscrow and PolicyManager events (by using old versions of the contracts). We don't implicitly generate those events anymore as part of the testing suite, so testing was skipped.

It's probably fair to remove the tests - I don't see a simple way to get them working. Not in this PR, but we might need to think about just removing the `nucypher status events` functionality. It's only been kept for legacy reasons, which become less needed/applicable over time.

File: nucypher/cli/commands/ursula.py (hunk 12 above)
Comment (KPrasch) [reply nucypher/cli/commands/ursula.py:175]: Hmm, I suppose that's ambiguous usage.  If there's no value provided via `--operator-address` and `--force` is passed we'd need to crash to handle properly.  It looks like this is ore-existing behavior though so let's put it off into a follow-up PR?

Comment (derekpierre) [reply nucypher/cli/commands/ursula.py:175]: Sure 👍 , but let's get an issue filed before we merge - don't want to lose track of it.

File: tests/utils/mock.py [hunk 15]
```diff
@@ -0,0 +1,9 @@
+from nucypher.blockchain.economics import EconomicsFactory
```
Comment (derekpierre) [inline tests/utils/mock.py:1]: Is this file needed anymore?

**Review by derekpierre — DISMISSED** [review]
🎸  - minor comment about a file that can potentially be removed.

File: nucypher/blockchain/eth/actors.py [hunk 16]
```diff
@@ -288,14 +291,42 @@ class Operator(BaseActor):
...
-                 operator_address: ChecksumAddress = None,
-                 *args, **kwargs):
+    def __init__(
+        self,
+        is_me: bool,
+        eth_provider_uri: str,
+        payment_method: ContractPayment,
+        work_tracker: Optional[WorkTracker] = None,
+        operator_address: Optional[ChecksumAddress] = None,
+        signer: Signer = None,
```
Comment (jMyles) [inline nucypher/blockchain/eth/actors.py:301]: Looking at the complete PR diff in order, this is the first line that made me go, "huh, I wonder why we're going in this direction."

**Review by theref — COMMENTED** [review]
Searching for remaining references to `federated` shows that there's quite a bit in the examples, documentation, and CI:
 https://github.com/nucypher/nucypher/blob/64c81e9b6acf1764bc740b2b1bcf8b7764154ee7/deploy/services/ursula_node.j2#L8

https://github.com/nucypher/nucypher/blob/development/docs/source/support/contribution.rst

https://github.com/nucypher/nucypher/blob/development/docs/source/application_development/local_fleet_demo.rst

https://github.com/nucypher/nucypher/blob/development/scripts/ci/docker-compose.yml

https://github.com/nucypher/nucypher/blob/development/examples/finnegans_wake_demo/finnegans-wake-demo-federated.py

https://github.com/nucypher/nucypher/blob/development/scripts/local_fleet/run_lonely_ursula.py

I think that some of these files can just be removed completely

**Comment by KPrasch** [comment]
Thanks for pointing these out.  It appears I lose some deletions in the most recent rebase.  I'll parse the docs and scripts here and make the appropriate removals.

File: tests/utils/mock.py (hunk 15 above)
Comment (KPrasch) [reply tests/utils/mock.py]: 👍🏻

**Review by cygnusv — COMMENTED** [review]
Some preliminary comments, but looks great. Thanks for the cleanup!

File: nucypher/characters/lawful.py [hunk 17]
```diff
@@ -1164,27 +1097,16 @@ def to_json(self) -> Dict[str, Any]:
                     )
 
 
-class Enrico(Character):
-    """A Character that represents a Data Source that encrypts data for some policy's public key"""
+class Enrico:
```
[the remark this answers is not in the document]
Comment (cygnusv) [reply nucypher/characters/lawful.py:1100]: Agreed. In that case, then I think the change in the docstring and removing that method is in order.

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/lawful.py:1100]: Which method?

[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/characters/lawful.py:1100]: Related to this comment - https://github.com/nucypher/nucypher/pull/3030#discussion_r1047066164

File: nucypher/characters/lawful.py [hunk 18]
```diff
@@ -959,8 +904,6 @@ def from_seednode_metadata(cls, seednode_metadata, *args, **kwargs):
         """
         Essentially another deserialization method, but this one doesn't reconstruct a complete
         node from bytes; instead it's just enough to connect to and verify a node.
-
-        NOTE: This is a federated only method.
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/lawful.py:963]: > Although at first glance, it does appear that this method is used in non-federated setting...

The note that is removed here was inaccurate.   In fact this method is used as a core part of node verification logic during p2p comms.

File: nucypher/characters/lawful.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (cygnusv) [reply nucypher/characters/lawful.py:1100]: > Which method?

the method def _set_known_node_class

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/lawful.py:1100]: 👍🏻 I was confused since this thread is related to `Enrico`.

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/lawful.py:1100]: 👍🏻

File: nucypher/cli/actions/auth.py [hunk 19]
```diff
@@ -48,9 +45,9 @@ def unlock_signer_account(config: CharacterConfiguration, json_ipc: bool) -> Non
     else:
         account = config.checksum_address
 
-    eth_password_is_needed = all((not config.federated_only,
-                                  not config.signer.is_device(account=account),
-                                  not config.dev_mode))
+    eth_password_is_needed = all(
+        (not config.signer.is_device(account=account), not config.dev_mode)
+    )
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/actions/auth.py:50]: 👍🏻

File: nucypher/cli/commands/ursula.py (hunk 12 above)
Comment (KPrasch) [reply nucypher/cli/commands/ursula.py:175]: https://github.com/nucypher/nucypher/issues/3039

File: docs/source/support/contribution.rst [hunk 20]
```diff
@@ -387,4 +353,4 @@ For example, for a new ``patch`` release, we would do:
   (nucypher)$ make release bump=patch
 
 3. The previous step triggers the publication webhooks on CircleCI.
```
Comment (derekpierre) [inline docs/source/support/contribution.rst:355]: Do we still use CircleCI for release publication?

File: nucypher/utilities/prometheus/collector.py [hunk 21]
```diff
@@ -113,12 +113,11 @@ def _collect_internal(self) -> None:
         self.metrics["learning_status"].state('running' if self.ursula._learning_task.running else 'stopped')
         self.metrics["known_nodes_gauge"].set(len(self.ursula.known_nodes))
 
-        if not self.ursula.federated_only:
-            decentralized_payload = {
-                "staking_provider_address": self.ursula.checksum_address,
-                "operator_address": self.ursula.operator_address,
-            }
-            base_payload.update(decentralized_payload)
+        decentralized_payload = {
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/utilities/prometheus/collector.py:116]: +1

**Review by derekpierre — DISMISSED** [review]
🎸

File: nucypher/cli/actions/auth.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (jMyles) [reply nucypher/cli/actions/auth.py:50]: How about 

`eth_password_is_not_needed = config.signer.is_device(account=account) or config.dev_mode` ?

File: nucypher/blockchain/eth/actors.py [hunk 22]
```diff
@@ -288,14 +291,42 @@ class Operator(BaseActor):
...
+
+        if not transacting_power:
+            transacting_power = TransactingPower(
+                account=operator_address,
+                password=client_password,
+                signer=signer,
+                cache=True,
+            )
+        self.transacting_power = transacting_power
+        crypto_power.consume_power_up(transacting_power)
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py:323]: The Character/Power API vision is incompletely implemented: the transacting power follows the interface specification for CryptoPowerUp but it is not consumed via `ursula.crypto_powers`, instead transacting power is directly set here and later injected into agents.  I've removed the superfluous `crypto_power` init parameter ans subsequent call to `consume_power_up` to make this more clear.

**Review by jMyles — DISMISSED** [review]
I'm prepared to approve this - it is outstanding work, especially given the magnitude of this change.  However, I do think that it decreases fidelity of the over-arching interface with respect to Character instantiation - particularly the matter of which parameters go into which constructors and why.  This needs to be highly cognizable in order to avoid becoming a maintenance hazard.

Another contingency is the merge of https://github.com/KPrasch/nucypher/pull/41, which @KPrasch has already endorsed in spirit.

I also advocate turning our attention now to a new branch beginning with `d235b821` as the next order of operations (though it's possible we can work contemporaneously on #3031, now that we are shoring up the interface to MVP level).

**Comment by KPrasch** [comment]
> Another contingency is the merge of https://github.com/KPrasch/nucypher/pull/41, which @KPrasch has already endorsed in spirit.

merged KPrasch/nucypher/pull/41

File: nucypher/utilities/prometheus/collector.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/utilities/prometheus/collector.py:116]: 👍🏻

File: nucypher/cli/actions/auth.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/actions/auth.py:50]: 👍🏻 (removed the call to `all`)

File: docs/source/support/contribution.rst (hunk 20 above)
Comment (KPrasch) [reply docs/source/support/contribution.rst:355]: No... this appears to be some leftover reference to circleCI.  Let's make an update here alongside a package publication GH action instead of changing it in this PR.
